android/library/src/main/java/com/polidea/multiplatformbleadapter/BleModule.java

```diff
@@ -157,533 +127,793 @@ public void destroyClient() {
-        return getReactApplicationContext().getPackageManager().hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE);
-    }
+        final OneTimeActionExecutor<BleError> oneTimeErrorCallback = new OneTimeActionExecutor<BleError>() {
+            @Override
+            public void action(BleError error) {
+                onErrorCallback.onError(error);
+            }
+        };
 
-    @BluetoothState
-    private String nativeAdapterStateToReactNativeBluetoothState(int adapterState) {
-        switch (adapterState) {
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
+            final Subscription subscription = connection
+                    .requestMtu(mtu)
+                    .doOnUnsubscribe(new Action0() {
+                        @Override
+                        public void call() {
+                            oneTimeErrorCallback.execute(BleErrorUtils.cancelled());
+                            pendingTransactions.removeSubscription(transactionId);
+                        }
+                    }).subscribe(new Observer<Integer>() {
+                        @Override
+                        public void onCompleted() {
+                            pendingTransactions.removeSubscription(transactionId);
+                        }
 
-            case BluetoothAdapter.STATE_OFF:
-                return BluetoothState.POWERED_OFF;
-            case BluetoothAdapter.STATE_ON:
-                return BluetoothState.POWERED_ON;
-            case BluetoothAdapter.STATE_TURNING_OFF:
-                // fallthrough
-            case BluetoothAdapter.STATE_TURNING_ON:
-                return BluetoothState.RESETTING;
-            default:
-                return BluetoothState.UNKNOWN;
-        }
-    }
+                        @Override
+                        public void onError(Throwable error) {
+                            oneTimeErrorCallback.execute(errorConverter.toError(error));
+                            pendingTransactions.removeSubscription(transactionId);
+                        }
 
-    @BluetoothState
-    private String rxAndroidBleAdapterStateToReactNativeBluetoothState(RxBleAdapterStateObservable.BleAdapterState rxBleAdapterState) {
-        if (rxBleAdapterState == RxBleAdapterStateObservable.BleAdapterState.STATE_ON) {
-            return BluetoothState.POWERED_ON;
-        } else if (rxBleAdapterState == RxBleAdapterStateObservable.BleAdapterState.STATE_OFF) {
-            return BluetoothState.POWERED_OFF;
+                        @Override
+                        public void onNext(Integer mtu) {
+                            device.setMtu(mtu);
+                            onSuccessCallback.onSuccess(device);
+                        }
+                    });
+
+            pendingTransactions.replaceSubscription(transactionId, subscription);
         } else {
-            return BluetoothState.RESETTING;
+            onSuccessCallback.onSuccess(device);
```

Comment:
Why is success emitted without any attempt to complete the operation? Why isn't this condition checked earlier?

Reply:
I don't know, it was already implemented in this way in RN lib. 
I assume that requesting MTU is not available on devices with pre-lollipop androids.
I will submit an issue for checking system version at the very beginning of this method.

Reply:
I'd prefer moving the platform-support check above the declaration of the oneTimeErrorCallback with inverted condition and a comment.

```
if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
    // requesting MTU is supported from LOLLIPOP up
    onSuccessCallback.onSuccess(device);
}
```

It kinda blends in after whole subscription business. 

But my other point is - shouldn't it be considered an error? The operation could not succeed, so in my opinion it failed. @Cierpliwy do you agree?

Reply:
I'd prefer moving the platform-support check above the declaration of the oneTimeErrorCallback with inverted condition and a comment.

```
if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
    // requesting MTU is supported from LOLLIPOP up
    onSuccessCallback.onSuccess(device);
}
```

It kinda blends in after whole subscription business. 

But my other point is - shouldn't it be considered an error? The operation could not succeed, so in my opinion it failed. @Cierpliwy do you agree?

Reply:
I already submitted an issue for this. At this moment I'm resolving this conversation.